### PR TITLE
Project version will not present commit hash on a fixed version

### LINKED ISSUE #196: Version in the logs should not include git commit hash
## Summary
Currently, when we print the project's version, we print also the commit hash:
```
INFO: 2020/11/10 15:32:04 main.go:20: CAKC048 Kubernetes Authenticator Client v0.18.1-e58708b starting up...
```

Fixed versions should not include the git commit tag (i.e `v0.18.1`) and should include only the version.

Note: Once we fix this, we should also fix [this issue](https://github.com/cyberark/secrets-provider-for-k8s/issues/258) in the secrets-provider as it has a similar mechanism.

## Steps to Reproduce
1. run `docker pull cyberark/conjur-authn-k8s-client:0.18.1`
2. run `docker run cyberark/conjur-authn-k8s-client:0.18.1`

## Expected Results
The log shows the message `CAKC048 Kubernetes Authenticator Client v0.18.1 starting up...`

## Actual Results (including error logs, if applicable)
The log shows the message `CAKC048 Kubernetes Authenticator Client v0.18.1-e58708b starting up...`

## Reproducible
   * [x] Always 
   
## Version/Tag number
0.18.1

### PR DESCRIPTION
**Pull request by benoaviram**
### What does this PR do?
In case we are building the project from a commit with a fixed version tag we will not present the git commit hash in the log.
The change affects the CAKC048 message log.
The change also affects the publish scripts but only in order to use the same function to describe the commit to figure out if this is a fixed commit (for release) or not. 

### What ticket does this PR close?
Resolves #196 

#### Change log
- [x] The CHANGELOG has been updated, or
- [ ] This PR does not include user-facing changes and doesn't require a CHANGELOG update

#### Test coverage
- [ ] This PR includes new unit and integration tests to go with the code changes, or
- [x] The changes in this PR do not require tests

#### Documentation
- [ ] Docs (e.g. `README`s) were updated in this PR, and/or there is a follow-on issue to update docs, or
- [ ] This PR does not require updating any documentation

#### Manual tests
**If you are preparing for a release**, have you run the following manual tests to verify existing functionality continues to function as expected?
- [ ] Manually run [Kubernetes-Conjur demo](https://github.com/conjurdemos/kubernetes-conjur-demo) with a local authn-k8s client image build

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -20,6 +20,8 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
   [cyberark/conjur-authn-k8s-client#206](https://github.com/cyberark/conjur-authn-k8s-client/issues/206)
 - Improve the error message raised when the username doesn't include the `host/` prefix
   [cyberark/conjur-authn-k8s-client#212](https://github.com/cyberark/conjur-authn-k8s-client/pull/212)
+- The project's version will not present the git commit hash on a fixed versions
```
[the remark this answers is not in the document]
Comment (orenbm) [reply CHANGELOG.md:23]: @shulifink can you assist please?

File: Dockerfile [hunk 2]
```diff
@@ -7,6 +7,7 @@ ENV GOOS=linux \
 
 # this value changes in ./bin/build
 ARG TAG="dev"
```
Comment (orenbm) [inline Dockerfile:9]: where is `TAG` used? do we still need it?

File: bin/build_utils [hunk 3]
```diff
@@ -25,3 +25,25 @@ function gen_versions() {
     echo $version
   done
 }
+
+function is_release_commit() {
+
```
Comment (orenbm) [inline bin/build_utils:30]: this empty line can be removed

File: bin/build_utils [hunk 4]
```diff
@@ -25,3 +25,25 @@ function gen_versions() {
...
+  if [ "$git_description" != "v${VERSION}" ]
+  then
+    true
+  else
+    false
+  fi
+}
+
+function append_commit_hash_to_version() {
```
Comment (orenbm) [inline bin/build_utils:43]: looks like this function should be called `git_tag_suffix`

Comment (orenbm) [reply bin/build_utils:43]: also - it is better to have this function above the other one as it is using it. it will read beter

File: pkg/authenticator/version.go [hunk 5]
```diff
@@ -6,11 +6,11 @@ import "fmt"
...
+// VersionSuffix field denotes the specific build type for the client. It may
 // be replaced by compile-time variables if needed to provide the git
 // commit information in the final binary
-var Tag = "dev"
+var VersionSuffix = "Dev"
 
 // FullVersionName is the user-visible aggregation of version and tag
 // of this codebase
-var FullVersionName = fmt.Sprintf("%s-%s", Version, Tag)
+var FullVersionName = fmt.Sprintf("%s%s", Version, VersionSuffix)
```
Comment (orenbm) [inline pkg/authenticator/version.go:16]: i would call it `TagSuffix` and add a comment saying that in fixed versions we don't want the tag to be present.
Also, it should be `var TagSuffix = "-Dev"`.

File: Dockerfile [hunk 6]
```diff
@@ -7,6 +7,7 @@ ENV GOOS=linux \
...
+ARG VersionSuffix="pre-release"
```
Comment (orenbm) [inline Dockerfile:10]: 1. Why is it `pre-release` and not `dev`?
2. It should be all-caps
3. I would call it `TAG_SUFFIX` as i say here: https://github.com/cyberark/conjur-authn-k8s-client/pull/224/files#r591725710

**Review by orenbm — CHANGES_REQUESTED** [review]
left a few comments

File: CHANGELOG.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (shulifink) [reply CHANGELOG.md:23]: I don;t get the format of the change log. Shouldn't it show what changed? it in past tense, or "No longer does this..." or "changed this to this", or "Now you can this and that", or "Fixed this so that blah blah"?

[the remark this answers is not in the document]
Comment (shulifink) [reply CHANGELOG.md:23]: So eg  here I would say something like:

Doesn't include git commit hash in printed version for fixed versions

What/who is not including the git command hash?

[the remark this answers is not in the document]
Comment (benoaviram) [reply CHANGELOG.md:23]: @shulifink @orenbm  thanks for the comments
@shulifink during the build process (while creating the product container) the script checks the commits to identify whether this is a fixed version and in that case it will not concatenate the git commit hash to its version.
changing the description to your suggestion

File: Dockerfile (hunk 2 above)
Comment (benoaviram) [reply Dockerfile:9]: Its being used for tagging the container image. you are correct that we dont need it in the docker file.. deleted

File: bin/build_utils (hunk 4 above)
Comment (benoaviram) [reply bin/build_utils:43]: done

File: bin/build_utils [hunk 7]
```diff
@@ -25,3 +25,25 @@ function gen_versions() {
...
+  if  is_release_commit
+  then
+    echo "-$(git_tag)"
+  fi
```
[the remark this answers is not in the document]
Comment (benoaviram) [reply bin/build_utils:47]: It should get the last command which will be empty but to make it more readable i am adding a else section

File: pkg/authenticator/version.go (hunk 5 above)
Comment (benoaviram) [reply pkg/authenticator/version.go:16]: Changed accordingly

File: CHANGELOG.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (shulifink) [reply CHANGELOG.md]: I underdatnd. Thanks for the explanation and for pointing out to me that I can look in the original ticket to see what was changed. This helped me understand what I was reviewing :)

So another question: What is a  "fixed version" ? Is this "fixed" in terms of the version was not working and now it's been fixed, or is this "fixed" as in it is a "static" version?

File: CHANGELOG.md [hunk 8]
```diff
@@ -20,6 +20,8 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
   [cyberark/conjur-authn-k8s-client#206](https://github.com/cyberark/conjur-authn-k8s-client/issues/206)
 - Improve the error message raised when the username doesn't include the `host/` prefix
   [cyberark/conjur-authn-k8s-client#212](https://github.com/cyberark/conjur-authn-k8s-client/pull/212)
+- Doesn't include git commit hash in printed version for fixed versions
```
Comment (orenbm) [inline CHANGELOG.md:23]: Don't...

File: CHANGELOG.md [hunk 9]
```diff
@@ -20,6 +20,8 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
...
+  [cyberark/conjur-authn-k8s-client#196](https://github.com/cyberark/conjur-authn-k8s-client/pull/196)
```
Comment (orenbm) [inline CHANGELOG.md:24]: can we link to the issue and not the PR?

File: Dockerfile [hunk 10]
```diff
@@ -6,7 +6,7 @@ ENV GOOS=linux \
     CGO_ENABLED=1
 
 # this value changes in ./bin/build
```
Comment (orenbm) [inline Dockerfile:8]: can you also elaborate here on why we use a suffix and not just the word?

File: Dockerfile [hunk 11]
```diff
@@ -21,7 +21,7 @@ RUN go get -u github.com/jstemmer/go-junit-report && \
     go get github.com/smartystreets/goconvey
 
 RUN go build -a -installsuffix cgo \
-    -ldflags="-X github.com/cyberark/conjur-authn-k8s-client/pkg/authenticator.Tag=$TAG" \
+    -ldflags="-X 'github.com/cyberark/conjur-authn-k8s-client/pkg/authenticator.TagSuffix=$TAG_SUFFIX'" \
```
Comment (orenbm) [inline Dockerfile:24]: why did you add the surrounding `'`? do we need it?

File: bin/build_utils [hunk 12]
```diff
@@ -25,3 +25,25 @@ function gen_versions() {
     echo $version
   done
 }
+function git_tag_suffix() {
```
Comment (orenbm) [inline bin/build_utils:28]: please add an empty line before this function

File: bin/build_utils [hunk 13]
```diff
@@ -25,3 +25,25 @@ function gen_versions() {
...
+  if  is_tagged_commit
```
Comment (orenbm) [inline bin/build_utils:29]: looks like you have 2 spaces after the if

File: bin/build_utils [hunk 14]
```diff
@@ -25,3 +25,25 @@ function gen_versions() {
...
+  then
```
Comment (orenbm) [inline bin/build_utils:30]: can you change this to `if [ is_tagged_commit ]; then`?

File: bin/build_utils [hunk 15]
```diff
@@ -25,3 +25,25 @@ function gen_versions() {
...
+    echo "-$(git_tag)"
+  else
+    echo ""
+  fi
+}
+function is_tagged_commit() {
```
Comment (orenbm) [inline bin/build_utils:36]: please add an empty line before this function

File: bin/build_utils [hunk 16]
```diff
@@ -25,3 +25,25 @@ function gen_versions() {
...
+
+  # fetching tags is required for git_description to work
+  git fetch --tags
+  git_description=$(git describe)
+
+  if [ "$git_description" != "v${VERSION}" ]
+  then
```
Comment (orenbm) [inline bin/build_utils:43]: please change to `if [ "$git_description" != "v${VERSION}" ]; then`

File: CHANGELOG.md (hunk 8 above)
Comment (shulifink) [reply CHANGELOG.md:23]: For fixed versions, no longer prints the git commit hash in the version.

**Review by shulifink — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (benoaviram) [reply CHANGELOG.md]: @shulifink the later... it refers to a release version

File: CHANGELOG.md (hunk 9 above)
Comment (benoaviram) [reply CHANGELOG.md:24]: it was a mistake.. fixed

File: Dockerfile (hunk 11 above)
Comment (benoaviram) [reply Dockerfile:24]: due to the space that comes up in a non-fixed versions..

File: bin/build_utils (hunk 16 above)
Comment (benoaviram) [reply bin/build_utils:43]: done

File: bin/build_utils (hunk 15 above)
Comment (benoaviram) [reply bin/build_utils:36]: done

File: bin/build_utils (hunk 12 above)
Comment (benoaviram) [reply bin/build_utils:28]: done

File: Dockerfile (hunk 10 above)
Comment (benoaviram) [reply Dockerfile:8]: we have this description in the version.go file... i think that this is enough without adding a description in every reference. let me think if you think otherwise

File: bin/build_utils [hunk 17]
```diff
@@ -25,3 +25,24 @@ function gen_versions() {
...
+    echo ""
+  fi
+}
+
+function is_tagged_commit() {
+  # fetching tags is required for git_description to work
+  git fetch --tags
+  git_description=$(git describe)
+
+  if [ "$git_description" != "v${VERSION}" ]; then
```
Comment (izgeri) [inline bin/build_utils:42]: we are trying to move away from using this logic in any pipelines, since it can lead to race conditions for builds

publish steps should always happen in the Jenkinsfile with the condition
`when { tag "v*" }`

in tag pipelines (ie when this condition is met), there is a `$TAG_NAME` environment variable that has the tag version. rather than do the logic you have here, you can just do
```suggestion
  if [[ ! -z "${TAG_NAME:-}" ]]; then
```

Comment (izgeri) [reply bin/build_utils:42]: for more info see: https://github.com/conjurinc/docs/blob/master/reference/jenkins.md#jenkins-pipeline-github-triggers

File: bin/build_utils [hunk 18]
```diff
@@ -25,3 +25,24 @@ function gen_versions() {
     echo $version
   done
 }
+
+function git_tag_suffix() {
+  if [ $(is_tagged_commit) ]; then
+    echo "-$(git_tag)"
```
Comment (izgeri) [inline bin/build_utils:31]: given the change below, this logic just becomes
```suggestion
  iif [[ ! -z "${TAG_NAME:-}" ]]; then
    echo "-$(TAG_NAME)"
```

File: bin/publish [hunk 19]
```diff
@@ -19,14 +19,10 @@ readonly TAGS=(
...
 
-# fetching tags is required for git_description to work
-git fetch --tags
-git_description=$(git describe)
-
 # if it’s not a tagged commit, VERSION will have extra characters (i.e. -g666c4b2), so that commit won't be published
 # only when tag matches the VERSION, push VERSION and latest releases
 # and x and x.y releases
-if [ "$git_description" = "v${VERSION}" ]; then
+if is_tagged_commit then
```
Comment (izgeri) [inline bin/publish:25]:
```suggestion
if [[ ! -z "${TAG_NAME:-}" ]]; then
```

**Review by izgeri — CHANGES_REQUESTED** [review]
I think these changes are actually taking the publishing strategy for this project in the wrong direction - we should be moving, if anything, to tag-triggered builds

It would also be _really_  helpful in this PR to share what the outcome of this change is, e.g. running in XX case produces an image with tag YY - for all of the flows this project runs. It's not obvious to me what the change will ultimately mean (note that using tag-triggers helps with this - when you use the presence of `TAG_NAME` in your case logic, it means you can easily test locally by running `TAG_NAME=0.11.3 ./bin/build` and see what the tags look like)

File: bin/publish (hunk 19 above)
Comment (benoaviram) [reply bin/publish:25]: Hey @izgeri, 
I do understand your suggestion, but do you think that this is aligned with Jenkins's best practice there - 
"The Jenkinsfile should contain minimal logic, it should only be used for assigning scripts to nodes and collecting results. The bulk of the build/test/release logic should be contained in other scripts in the repo." ?

**Review by shulifink — CHANGES_REQUESTED** [review]

File: bin/publish (hunk 19 above)
Comment (izgeri) [reply bin/publish:25]: can you please explain how this does not align with that best practice?

the Jenkinsfile only contains the logic about _when_ steps should be run (eg on tags)

the rest of the logic is entirely in the scripts (like this one we are commenting on, `bin/publish`)

this is consistent with the best practices recommended by the infra team for our pipelines, and with the direction all of our other projects are moving. in addition, it addresses a known bug where sometimes pipelines wouldn't publish because new commits have been added to master since the tag.

File: bin/publish [hunk 20]
```diff
@@ -19,14 +19,10 @@ readonly TAGS=(
...
-
-# if it’s not a tagged commit, VERSION will have extra characters (i.e. -g666c4b2), so that commit won't be published
-# only when tag matches the VERSION, push VERSION and latest releases
-# and x and x.y releases
-if [ "$git_description" = "v${VERSION}" ]; then
+# The push should run only on fixed versions, identified by the existence of the TAG_NAME environment variable
+# which happens automatically in tag-triggered builds. 
+# push VERSION and latest releases and x and x.y releases
+if [[ ! -z "${TAG_NAME:-}" ]]; then
   echo "Revision $git_description matches version $VERSION exactly"
```
[the remark this answers is not in the document]
Comment (benoaviram) [reply bin/publish]: change as part of the latest change to move the condition to the jenkinsfile

File: bin/build [hunk 21]
```diff
@@ -8,7 +8,7 @@ set -euo pipefail
 # (env GOOS=darwin GOARCH=amd64 go build) of project binaries
 IMAGE_NAME=conjur-authn-k8s-client
 TAG=dev
-VERSION="$(short_version_tag)"
+VERSION="${TAG_NAME//"v"}"
```
Comment (izgeri) [inline bin/build]: this should probably be conditional, so that it's TAG_NAME if the value is set, and short_version_name if not.

it's not critical, but it means in non-tag pipelines the RH image will have empty `version` and `release` labels, up to you if you change it.

**Review by izgeri — CHANGES_REQUESTED** [review]
Notes on trying this out locally:

## Running with `TAG_NAME` set

- Running `TAG_NAME=1.1.2 ./bin/build` builds local `conjur-authn-k8s-client` and `conjur-authn-k8s-client-redhat` images with tag `dev`
- Running `REDHAT_API_KEY=oops TAG_NAME=1.1.2 ./bin/publish`(with actual `docker push` commands commented) attempts to publish:
   ```
   This script should not run in a tag-triggered build. controlled by the JenkinsFile
   Pushing to Dockerhub...
   Tagging and pushing cyberark/conjur-kubernetes-authenticator:1.1.2
   Tagging and pushing cyberark/conjur-kubernetes-authenticator:latest
   Tagging and pushing cyberark/conjur-kubernetes-authenticator:1.1
   Tagging and pushing cyberark/conjur-kubernetes-authenticator:1
   Tagging and pushing cyberark/conjur-authn-k8s-client:1.1.2
   Tagging and pushing cyberark/conjur-authn-k8s-client:latest
   Tagging and pushing cyberark/conjur-authn-k8s-client:1.1
   Tagging and pushing cyberark/conjur-authn-k8s-client:1
   Pushing to RedHat container registry...
   ```
   
## Running without `TAG_NAME` set

- Running `./bin/build` builds local `conjur-authn-k8s-client` and `conjur-authn-k8s-client-redhat` images with tag `dev`
- Running `REDHAT_API_KEY=oops ./bin/publish`(with actual `docker push` commands commented) attempts to publish:
   ```
   $ REDHAT_API_KEY=oops ./bin/publish
   This script should not run in a tag-triggered build. controlled by the JenkinsFile
   Pushing to Dockerhub...
   Tagging and pushing cyberark/conjur-kubernetes-authenticator:
   Error parsing reference: "cyberark/conjur-kubernetes-authenticator:" is not a valid repository/tag: invalid reference format
   ```
   
   To fix this last flow, I think we should make the change I suggested in the review (exit with non-zero exit code when `TAG_NAME` is not set).

File: bin/build (hunk 21 above)
Comment (benoaviram) [reply bin/build]: it was changed in the second commit, I think it is better to have it simple, use the short_version_tag

File: bin/build_utils [hunk 22]
```diff
@@ -25,3 +21,17 @@ function gen_versions() {
...
+  if [[ -z "${TAG_NAME:-}" ]]; then
+    echo "-$(git_commit_short)"
+  else
+    echo ""
+  fi
+}
+
+function git_commit_short() {
+  git rev-parse --short HEAD
+}
```
Comment (orenbm) [inline bin/build_utils:37]: missing new line at EOF

File: bin/publish [hunk 23]
```diff
@@ -19,38 +19,29 @@ readonly TAGS=(
...
+  # you can't push the same tag twice to redhat registry, so ignore errors
+  if ! docker push "${REDHAT_IMAGE}:${VERSION}"; then
+    echo 'RedHat push FAILED! (maybe the image was pushed already?)'
+    exit 0
   fi
-fi
+else
+  echo 'Failed to log in to scan.connect.redhat.com'
+  exit 1
+fi
```
Comment (orenbm) [inline bin/publish:47]: missing new line at EOF

File: CHANGELOG.md [hunk 24]
```diff
@@ -20,6 +20,8 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
   [cyberark/conjur-authn-k8s-client#206](https://github.com/cyberark/conjur-authn-k8s-client/issues/206)
 - Improve the error message raised when the username doesn't include the `host/` prefix
   [cyberark/conjur-authn-k8s-client#212](https://github.com/cyberark/conjur-authn-k8s-client/pull/212)
+- The git commit hash is no longer included in the version number of a fixed version.
```
Comment (izgeri) [inline CHANGELOG.md:23]: Can this be more specific about the impact to the end user of this change? The Docker / RH images before only had the tag version - is what's changing the `version` label in the RH image, or something else? If this doesn't impact the end users, we don't need an entry here.

File: bin/publish [hunk 25]
```diff
@@ -19,34 +19,29 @@ readonly TAGS=(
...
-
-  echo "Pushing to Dockerhub..."
-  source_image=conjur-authn-k8s-client:dev
-  for destination_image_name in "${DESTINATION_IMAGES[@]}"; do
-    for tag in "${TAGS[@]}" $(gen_versions "$VERSION"); do
-      echo "Tagging and pushing $REGISTRY/$destination_image_name:$tag"
-      docker tag $source_image "$REGISTRY/$destination_image_name:$tag"
-      docker push "$REGISTRY/$destination_image_name:$tag"
-    done
+echo "This script should not run in a tag-triggered build. controlled by the JenkinsFile"
```
[the remark this answers is not in the document]
Comment (benoaviram) [reply bin/publish:22]: regarding the image's names - shouldn't changed with this PR. 
The only change that is being made here is in the CAKC048 log - as mentioned in the issue's details, it should represent the commit's hash in fixed versions which is based on the existence of the TAG_NAME environment variable which is being set automatically by the tag-triggered builds. 

Regarding the publish script, fixing it.

File: CHANGELOG.md (hunk 24 above)
Comment (benoaviram) [reply CHANGELOG.md:23]: as mentioned above, the affect on the end user is only in the log message - 
Before this PR it represent the commit's hash all the times - 
 CAKC048 Kubernetes Authenticator Client **v0.18.1-e58708b** starting up...

This PR changes it, in case of a fixed version it will be - 
 CAKC048 Kubernetes Authenticator Client **v0.18.1** starting up...

And in case of a non-fixed version it will be as before 
 CAKC048 Kubernetes Authenticator Client **v0.18.1-e58708b** starting up...

so i think that description is correct, agree?

Comment (izgeri) [reply CHANGELOG.md:23]: it's not clear to me - I might revise it to something like
```suggestion
- The `CAKC048` log message now shows the release version for release builds
   and no longer includes the git commit hash in the log output.
```

File: Jenkinsfile [hunk 26]
```diff
@@ -62,7 +62,10 @@ pipeline {
 
     stage('Publish client Docker image') {
       when {
-        branch 'master'
+        allOf {
+          branch 'master'
+          tag "v*"
```
Comment (orenbm) [inline Jenkinsfile:67]: @benoaviram when will this happen? what does it mean?

File: bin/build_utils [hunk 27]
```diff
@@ -25,3 +21,17 @@ function gen_versions() {
     echo $version
   done
 }
+
+# returns the git commit's hash only for non-fixed versions, empty otherwise.
+# which can be identified by the TAG_NAME variable, which is being set automatically on tag-triggered builds
```
Comment (orenbm) [inline bin/build_utils:26]: i would change this sentence to `The TAG_NAME variable is set automatically on tag-triggered builds for non-fixed versions`

File: bin/publish [hunk 28]
```diff
@@ -5,7 +5,7 @@
 REDHAT_IMAGE='scan.connect.redhat.com/ospid-1c46a2de-1d88-40e6-a433-7114ad0099cb/conjur-openshift-authenticator-client'
 
 readonly REGISTRY="cyberark"
-readonly VERSION="$(short_version_tag)"
+readonly VERSION="${TAG_NAME//"v"}"
```
Comment (orenbm) [inline bin/publish:8]: can you explain this please?

File: bin/build_utils (hunk 27 above)
Comment (benoaviram) [reply bin/build_utils:26]: Actually this is the opposite, when pushing a new tag for a fixed version, a new tag-trigged build starts, which has the TAG_NAME env variable.

Comment (benoaviram) [reply bin/build_utils:26]: let me know if you think It's clear thats way

File: bin/publish (hunk 28 above)
Comment (benoaviram) [reply bin/publish:8]: We want to align this repo with other pipeline's standards - using the TAG_NAME variable which has the version in its value - just removing the "v" prefix in order to have the number's version

File: bin/build_utils (hunk 27 above)
Comment (orenbm) [reply bin/build_utils:26]: yes, please update the comment to say that

File: bin/publish (hunk 28 above)
Comment (orenbm) [reply bin/publish:8]: can you add a comment saying that we take the `TAG_NAME` variable and remove the `v` prefix? i didn't know what this syntax is

File: Jenkinsfile (hunk 26 above)
Comment (benoaviram) [reply Jenkinsfile:67]: Just taking this condition from the publish script to the jenkins file. 
In the past we had this condition in the jenkins file which ran the publish script only on master branch and inside the publish script there was a check for a fixed-version (using the old way - git describe).
From now - the condition resides on the Jenkins file, only on master master branch which has tag (fixed-version) we will run the publish script

File: CHANGELOG.md (hunk 24 above)
Comment (benoaviram) [reply CHANGELOG.md:23]: fixed

**Review by orenbm — DISMISSED** [review]

**Review by orenbm — APPROVED** [review]

**Review by izgeri — APPROVED** [review]
LGTM! Summary of what I verified below.

## Running with `TAG_NAME` set

- Running `TAG_NAME=1.1.2 ./bin/build` builds local `conjur-authn-k8s-client` and `conjur-authn-k8s-client-redhat` images with tag `dev`

   The logs show:
   ```
   $ docker run conjur-authn-k8s-client:dev
   INFO:  2021/03/22 14:25:09.745555 main.go:16: CAKC048 Kubernetes Authenticator Client v0.19.1 starting up...
   ```

- Running `REDHAT_API_KEY=oops TAG_NAME=1.1.2 ./bin/publish`(with actual `docker push` commands commented) attempts to publish:
   ```
   Pushing to Dockerhub...
   Tagging and pushing cyberark/conjur-kubernetes-authenticator:1.1.2
   Tagging and pushing cyberark/conjur-kubernetes-authenticator:latest
   Tagging and pushing cyberark/conjur-kubernetes-authenticator:1.1
   Tagging and pushing cyberark/conjur-kubernetes-authenticator:1
   Tagging and pushing cyberark/conjur-authn-k8s-client:1.1.2
   Tagging and pushing cyberark/conjur-authn-k8s-client:latest
   Tagging and pushing cyberark/conjur-authn-k8s-client:1.1
   Tagging and pushing cyberark/conjur-authn-k8s-client:1
   Pushing to RedHat container registry...
   ```
   
## Running without `TAG_NAME` set

- Running `./bin/build` builds local `conjur-authn-k8s-client` and `conjur-authn-k8s-client-redhat` images with tag `dev`

   The logs show:
   ```
   $ docker run conjur-authn-k8s-client:dev
   INFO:  2021/03/22 14:43:13.922822 main.go:16: CAKC048 Kubernetes Authenticator Client v0.19.1-c336b82 starting up...
   ```

- Running `REDHAT_API_KEY=oops ./bin/publish`(with actual `docker push` commands commented) does not attempt to publish, and prints a warning:
   ```
   $ REDHAT_API_KEY=oops ./bin/publish
   This script should only run in a tag-triggered build controlled by the JenkinsFile
   ```

File: bin/publish [hunk 29]
```diff
@@ -19,38 +20,32 @@ readonly TAGS=(
...
-      echo 'RedHat push FAILED! (maybe the image was pushed already?)'
-      exit 0
-    fi
-  else
-    echo 'Failed to log in to scan.connect.redhat.com'
-    exit 1
+if docker login scan.connect.redhat.com -u unused -p "$REDHAT_API_KEY"; then
+  # you can't push the same tag twice to redhat registry, so ignore errors
+  if ! docker push "${REDHAT_IMAGE}:${VERSION}"; then
+    echo 'RedHat push FAILED! (maybe the image was pushed already?)'
```
Comment (shulifink) [inline bin/publish:45]: Red Hat push failed. The image might have been pushed already.

**Review by shulifink — COMMENTED** [review]
@abrahamko  - small things....(sorry so delayed)

File: bin/publish [hunk 30]
```diff
@@ -19,38 +20,32 @@ readonly TAGS=(
...
-  echo "Pushing to Dockerhub..."
-  source_image=conjur-authn-k8s-client:dev
-  for destination_image_name in "${DESTINATION_IMAGES[@]}"; do
-    for tag in "${TAGS[@]}" $(gen_versions "$VERSION"); do
-      echo "Tagging and pushing $REGISTRY/$destination_image_name:$tag"
-      docker tag $source_image "$REGISTRY/$destination_image_name:$tag"
-      docker push "$REGISTRY/$destination_image_name:$tag"
-    done
+if [[ -z "${TAG_NAME:-}" ]]; then
+  echo "This script should only run in a tag-triggered build controlled by the JenkinsFile"
```
Comment (shulifink) [inline bin/publish:24]: should run, or may run (ie allowed to run) ?

File: bin/publish [hunk 31]
```diff
@@ -19,38 +20,32 @@ readonly TAGS=(
...
+  exit 1
+fi
+
+echo "Pushing to Dockerhub..."
```
Comment (shulifink) [inline bin/publish:28]: Docker Hub (according to their site)
